File: ImageEffectBackend/Libraries/RotationLibrary/Rotation.cpp
#include "Rotation.h"
using namespace std;

vector<vector<Pixel>> Rotation::applyRotation(vector<vector<Pixel>>& image, int value) {
    //applyRotation method in Rotation class
    vector<vector<Pixel>> resultImage(image[0].size(), vector<Pixel>(image.size()));
    //result image vector

    //checking the cases one-by-one

    if (value == 0) {
        //case where there is no rotation
        return image;
    }
    else if (value == 1) {
        //case where the rotation is by 90 degrees in clock-wise direction
        for (int i = 0; i < image.size(); i++) {
            for (int j = 0; j < image[i].size(); j++) {
                resultImage[j][image.size() - 1 - i] = image[i][j];
            }
        }
    }
    else if (value == 2) {
        //case where the rotation is by 180 degrees in clock-wise direction
        vector<vector<Pixel>>test;//temporary vector
        for(int i = image.size()-1;i>=0;i--){
            test.push_back(image[i]);
        }
        resultImage = test;
    }
    else if (value == 3) {
        //case where the rotation is by 270 degrees in clock-wise direction
        for (int i = 0; i < image.size(); i++) {
            for (int j = 0; j < image[i].size(); j++) {
                resultImage[image[i].size() - 1 - j][i] = image[i][j];
            }
        }
    }
    else {
        return image;
    }

    return resultImage;//returning the modified image vector
}

File: ImageEffectBackend/src/main/java/com/iiitb/imageEffectApplication/effectImplementations/Rotation.java
package com.iiitb.imageEffectApplication.effectImplementations;
import com.iiitb.imageEffectApplication.baseEffects.DiscreteEffect;
import com.iiitb.imageEffectApplication.baseEffects.SingleValueDiscreteEffect;
import com.iiitb.imageEffectApplication.exception.IllegalParameterException;
import com.iiitb.imageEffectApplication.service.LoggingService;
import libraryInterfaces.Pixel;
import static libraryInterfaces.RotationInterface.applyRotation;

/*Rotation class implements SingleValueDiscreteEffect and to apply rotation effect. */

public class Rotation implements SingleValueDiscreteEffect {
    private int parameterValue;// Parameter to store the rotation value

    public Rotation(){}//Default constructor
    public Pixel[][] apply(Pixel[][] image, String fileName, LoggingService loggingService){
        //This method applies the rotation effect on the given image based on the given parameter value.
        Pixel[][] resultImage = applyRotation(image,this.parameterValue);
        String optionValues = "";
        if(this.parameterValue == 0){
            optionValues = "No Rotation";
        }
        if(this.parameterValue == 1){
            optionValues = "Rotation by 90 degrees in clockwise direction";
        }
        if(this.parameterValue == 2){
            optionValues = "Rotation by 180 degrees in clockwise direction";
        }
        if(this.parameterValue == 3){
            optionValues = "Rotation by 270 degrees in clockwise direction";
        }
        loggingService.addLog(fileName,"Rotation",optionValues);// Add a log entry for the applied rotation effect.
        return resultImage;// Return the result image.
    }

    public void setParameterValue(int parameterValue) throws IllegalParameterException{
        /*This method sets the rotation parameter value, validating that it falls within a valid range.
        checking if the brightness parameter value is within range of 0 to 3*/
        if (parameterValue < 0 || parameterValue > 3) {
            throw new IllegalParameterException("Not a valid option");
        }
        this.parameterValue = parameterValue;//Setting the parameter value
    }
}
